src/nnet3/discriminative-supervision.cc

```diff
@@ -0,0 +1,450 @@
+      // throw an exception. 
+      KALDI_ERR << "Error reading Lattice from stream";
+    }
+    den_lat = *lat;
+    delete lat;
+    TopSort(&den_lat);
+  }
+
+  ExpectToken(is, binary, "</DiscriminativeSupervision>");
+}
+
+bool LatticeToDiscriminativeSupervision(const std::vector<int32> &num_ali,
+                                        const Lattice &den_lat, 
+                                        BaseFloat weight,
+                                        DiscriminativeSupervision *supervision) {
+  supervision->weight = weight;
+  supervision->num_sequences = 1;
+  supervision->frames_per_sequence = num_ali.size();
+  supervision->num_ali = num_ali;
+  supervision->den_lat = den_lat;
+  TopSort(&(supervision->den_lat));
+
+  supervision->Check();
+
+  return true;
+}
+
+void DiscriminativeSupervision::Check() const {
+  int32 num_frames_subsampled = num_ali.size();
+
+  {
+    std::vector<int32> state_times;
+    int32 max_time = LatticeStateTimes(den_lat, &state_times);
+    KALDI_ASSERT(max_time == num_frames_subsampled);
+  }
+}
+
+DiscriminativeSupervisionSplitter::DiscriminativeSupervisionSplitter(
+    const SplitDiscriminativeSupervisionOptions &config,
+    const TransitionModel &tmodel,
+    const DiscriminativeSupervision &supervision):
+    config_(config), tmodel_(tmodel), supervision_(supervision) {
+  if (supervision_.num_sequences != 1) {
+    KALDI_WARN << "Splitting already-reattached sequence (only expected in "
+               << "testing code)";
+  }
+
+  KALDI_ASSERT(supervision_.num_sequences == 1); // For now, don't allow splitting already merged examples
+
+  den_lat_ = supervision_.den_lat;
+  PrepareLattice(&den_lat_, &den_lat_scores_);
+  
+  int32 num_states = den_lat_.NumStates(),
+        num_frames = supervision_.frames_per_sequence * supervision_.num_sequences;
+  KALDI_ASSERT(num_states > 0);
+  int32 start_state = den_lat_.Start();
+  // Lattice should be top-sorted and connected, so start-state must be 0.
+  KALDI_ASSERT(start_state == 0 && "Expecting start-state to be 0");
+  
+  KALDI_ASSERT(num_states == den_lat_scores_.state_times.size());
+  KALDI_ASSERT(den_lat_scores_.state_times[start_state] == 0);
+  KALDI_ASSERT(den_lat_scores_.state_times.back() == num_frames);
+}
+
+// Make sure that for any given pdf-id and any given frame, the den-lat has
+// only one transition-id mapping to that pdf-id, on the same frame.
+// It helps us to more completely minimize the lattice.  Note: we
+// can't do this if the criterion is MPFE, because in that case the
+// objective function will be affected by the phone-identities being
+// different even if the pdf-ids are the same.
+void DiscriminativeSupervisionSplitter::CollapseTransitionIds(
+    const std::vector<int32> &state_times, Lattice *lat) const {
+  typedef Lattice::StateId StateId;
+  typedef Lattice::Arc Arc;
+
+  int32 num_frames = state_times.back();   // TODO: Check if this is always true
```

Comment:
it's always true since it's a non-compact lattice, and you've ensured that it's topologically sorted.